Lib/test/test_shelve.py

```diff
@@ -165,6 +168,277 @@ def test_default_protocol(self):
+                    self.fn,
+                    protocol=proto,
+                    serializer=serializer,
+                    deserializer=deserializer,
+                ) as s:
+                    bar = "bar"
+                    bytes_data = b"Hello, world!"
+                    bytearray_data = bytearray(b"\x00\x01\x02\x03\x04")
+                    array_data = array.array("i", [1, 2, 3, 4, 5])
+
+                    s["foo"] = bar
+                    s["bytes_data"] = bytes_data
+                    s["bytearray_data"] = bytearray_data
+                    s["array_data"] = array_data
+
+                    if proto == 5:
+                        self.assertEqual(s["foo"], str(bar))
+                        self.assertEqual(s["bytes_data"], "Hello, world!")
+                        self.assertEqual(
+                            s["bytearray_data"], bytearray_data.decode()
+                        )
+                        self.assertEqual(
+                            s["array_data"], array_data.tobytes().decode()
+                        )
+                    else:
+                        self.assertEqual(s["foo"], "str")
+                        self.assertEqual(s["bytes_data"], "bytes")
+                        self.assertEqual(s["bytearray_data"], "bytearray")
+                        self.assertEqual(
+                            s["array_data"], array_data.tobytes().decode()
+                        )
+
+        with self.assertRaises(AssertionError):
+            def serializer(obj, protocol=None):
+                return bytes(type(obj).__name__, 'utf-8')
+
+            def deserializer(data):
+                pass
+
+            with shelve.open(self.fn,
+                             serializer=serializer,
+                             deserializer=deserializer) as s:
+                s["foo"] = "bar"
+                self.assertEqual(s["foo"], "bar")
+
+    def test_custom_incomplete_serializer_and_deserializer(self):
+        dbm_sqlite3 = import_helper.import_module("dbm.sqlite3")
+        os.mkdir(self.dirname)
+        self.addCleanup(os_helper.rmtree, self.dirname)
+
+        with self.assertRaises(dbm_sqlite3.error):
+            def serializer(obj, protocol=None):
+                pass
+
+            def deserializer(data):
+                return BytesIO(data).read().decode("utf-8")
+
+            with shelve.open(self.fn,
+                             serializer=serializer,
+                             deserializer=deserializer) as s:
+                s["foo"] = "bar"
+                self.assertEqual(s["foo"], "bar")
+
+    def test_custom_serializer_and_deserializer_bsd_db_shelf(self):
+        berkeleydb = import_helper.import_module("berkeleydb")
+
+        def serializer(obj, protocol=None):
+            if protocol == 5:
+                return bytes(f"{len(type(obj).__name__)}", encoding="utf-8")
+            else:
+                return bytes(f"{type(obj).__name__}", "utf-8")
```

Comment:
Could I suggest that you use `marshal` for the tests instead of those weird serialize/deserialize? (I don't know about berkeleydb). If you don't want to
you replace this code by:

```py
data = obj.__class__.__name__
if protocol == 5:
	data = str(len(data))
return data.encode("utf-8")
```

which is a bit easier to read IMO.